Feature predefined response

atlantees opened the pull request.

atlantees committed:
Created modal for adding keywords and predefined responses

atlantees committed:
adjusted some text-size

eirism committed:
Add backend-part of predefined responses

Keywords are extracted by an external API.
After the keyword has been extracted, if is searched for a matching
response in the DB. The response is sent as the third element in
the question response socketIO event (null if none was found).
The matching response is also saved together with the question.
There is also a listener for socketIO for adding new keyword-responses
The extra flags in requirements.txt are to be able to
build pillow without extra dependecies.

eirism committed:
Merge remote-tracking branch 'origin/feature-predefinedResponses' into feature-predefined-response

atlantees committed:
Fixed some old-school codestyle

atlantees committed:
Added function to send keyword/response-tuples from lecturer_session

atlantees committed:
Created init for Response-records

atlantees committed:
Check if a question has the attribute predefined response and displays it

atlantees committed:
Styling, if you can call it that

atlantees committed:
Reloads pages to display correct responses on new_keywords event

atlantees committed:
Documentation, kinda

atlantees committed:
Adds live update of responses throughout all sites

atlantees committed:
Merge branch 'development' into feature-predefined-response

atlantees committed:
Fixed comment

eirism committed:
Hopefully fix requirements

Travis seems to be able to install pillow from binary

atlantees committed:
Fixes to codestyle

eirism commented on iris/views.py:
```diff
@@ -241,6 +287,11 @@ def get_lecture_session(course_id):
 
 
+def get_question(session_id, question):
+    return get_model_or_create(models.Questions, {'session_id': session_id,
```
This gives an error if there are more than one question with the same text.

eirism commented on iris/views.py:
```diff
@@ -118,6 +127,43 @@ def handle_feedback(message, l_session, course_id):
+    for keyword in keywords.split(','):
+        keyword = keyword.strip()
+        new_keyword = models.Response(keyword, course_id, response)
+        db.session.add(new_keyword)
```
Should check if the keyword exists. Currently it crashes if it exists, maybe we should update the response text?

eirism commented on iris/static/js/lecturer_session.js:
```diff
@@ -128,3 +150,36 @@ let difficulty = new Chart(ctxDifficulty, {
+let dialog = document.querySelector('dialog')
+let showDialogButton = document.querySelector('#show-dialog')
+if (!dialog.showModal) {
+  dialogPolyfill.registerDialog(dialog)
```
dialogPolyfill is not defined in Firefox, and this breaks the modal.

atlantees committed:
Gets question based on PK instead of a useless combination of fields

atlantees committed:
Adding an already existing keyword will update the associated response

eirism committed:
Fix dialog polyfill in Firefox

Uses the dialog-polyfill project by Google Chrome.
The dialog tag was moved to be a child of body to prevent
it being displayed inside the parent container,
https://github.com/GoogleChrome/dialog-polyfill/#stacking-context

atlantees committed:
made some nice changes to codestyle
